src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs

```diff
@@ -806,6 +862,62 @@ public TResult Result
+            WaitAsync(timeout, default);
+
+        /// <summary>Gets a <see cref="ValueTask{TResult}"/> that will complete when this <see cref="ValueTask{TResult}"/> completes, when the specified timeout expires, or when the specified <see cref="CancellationToken"/> has cancellation requested.</summary>
+        /// <param name="timeout">The timeout after which the <see cref="ValueTask{TResult}"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Canceling or timing out the <see cref="WaitAsync"/> doesn't affect the execution of the original task itself; if a <see cref="WaitAsync"/> is canceled or times out on a task that
+        /// hasn't yet completed, the original task may continue to execute even after the <see cref="WaitAsync"/> task has completed.
+        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask{TResult}"/> shouldn't be accessed again, regardless of
+        /// whether the operation completes, times out, or is canceled.
+        /// </remarks>
+        public ValueTask<TResult> WaitAsync(TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            uint millisecondsTimeout = Task.ValidateTimeout(timeout, ExceptionArgument.timeout);
+
+            if (IsCompleted)
+            {
+                return this;
+            }
+
+            // Perf note: If the ValueTask represents an incomplete operation backed by an IValueTaskSource, this implementation will result in
+            // one more Task allocation than is truly necessary, as it first marshals the IValueTaskSource to a Task via AsTask(), and then uses
+            // the Task.WaitAsync method on that.  It would be possible to instead have a specialized AsTask(timeout, cancellationToken)
+            // routine that combined all of that logic into one, saving a Task instance, but at present that doesn't seem worth all the extra
+            // code and complexity required.  This can be revisited later should it prove impactful.  We could also preempt the AsTask() by checking
+            // for a timeout of 0 or an already-canceled token, but both cases already represent expensive code paths that will trigger exceptions,
+            // so it's also not worth the extra code at present.  If we did decide to change something here, we would need to ensure that
+            // even when timing out or canceling, the IValueTaskSource eventually has GetResult called on it, in order to enable the producer
+            // of that instance to be notified that it's no longer in use (e.g. for pooling); AsTask() handles that for us implicitly.
```

Comment:
very minor nit: extra spaces

```suggestion
            // Perf note: If the ValueTask represents an incomplete operation backed by an IValueTaskSource, this implementation will result in
            // one more Task allocation than is truly necessary, as it first marshals the IValueTaskSource to a Task via AsTask(), and then uses
            // the Task.WaitAsync method on that. It would be possible to instead have a specialized AsTask(timeout, cancellationToken)
            // routine that combined all of that logic into one, saving a Task instance, but at present that doesn't seem worth all the extra
            // code and complexity required. This can be revisited later should it prove impactful. We could also preempt the AsTask() by checking
            // for a timeout of 0 or an already-canceled token, but both cases already represent expensive code paths that will trigger exceptions,
            // so it's also not worth the extra code at present. If we did decide to change something here, we would need to ensure that
            // even when timing out or canceling, the IValueTaskSource eventually has GetResult called on it, in order to enable the producer
            // of that instance to be notified that it's no longer in use (e.g. for pooling); AsTask() handles that for us implicitly.
```

Reply:
I almost always use two spaces after a period :)